Issue #166 - Replica.updateMetastore calls metastore with batched partition lists.

barnharts4 opened the pull request:
Fixes Issue #166 

Log file output from running test on table with 5019 partitions:
```
21:01:53.209 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Updating replica table metadata.
21:01:53.334 [main] INFO  com.hotels.hcommon.hive.metastore.client.closeable.CloseableMetaStoreClientInvocationHandler - Couldn't invoke method public abstract org.apache.hadoop.hive.metastore.api.Table org.apache.hadoop.hive.metastore.IMetaStoreClient.getTable(java.lang.String,java.lang.String) throws org.apache.hadoop.hive.metastore.api.MetaException,org.apache.thrift.TException,org.apache.hadoop.hive.metastore.api.NoSuchObjectException
21:11:02.672 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating 5019 new partitions.
21:11:02.677 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 0 through 999
21:15:21.485 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 1000 through 1999
21:19:35.692 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 2000 through 2999
21:23:57.927 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 3000 through 3999
21:28:13.187 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 4000 through 4999
21:32:30.142 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 5000 through 5018
21:32:35.818 [main] INFO  hive.metastore - Closed a connection to metastore, current connections: 0
21:32:35.818 [main] INFO  com.hotels.bdp.circustrain.core.PartitionedTableReplication - Replicated 5019 partitions of table test.barnparttest5000.
```

abhimanyugupta07 commented on circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java:
```diff
@@ -189,7 +222,13 @@ public void updateMetadata(
+          for (List<Partition> sublist : Lists.partition(partitionsToAlter, partitionBatchSize)) {
+            int start = counter * partitionBatchSize;
+            LOG.info("Altering partitions {} through {}", start, start + sublist.size() - 1);
+            client.alter_partitions(replicaDatabaseName, replicaTableName, partitionsToAlter);
```
shouldn't this be `sublist` instead of `partitionsToAlter`.

`client.alter_partitions(replicaDatabaseName, replicaTableName, partitionsToAlter);`

barnharts4 replied:
good catch, thanks!

abhimanyugupta07 commented on circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java:
```diff
@@ -203,7 +242,13 @@ public void updateMetadata(
+          for (List<ColumnStatistics> sublist : Lists.partition(statisticsToSet, partitionBatchSize)) {
+            int start = counter * partitionBatchSize;
+            LOG.info("Setting column statistics for partitions {} through {}", start, start + sublist.size() - 1);
+            client.setPartitionColumnStatistics(new SetPartitionsStatsRequest(statisticsToSet));
```
here too. `sublist` instead of `statisticsToSet` ?

barnharts4 replied:
yep, thanks

abhimanyugupta07 commented on circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java:
```diff
@@ -425,18 +440,28 @@ public void alteringExistingPartitionedReplicaTableWithPartitionsSucceeds() thro
+    assertThat(firstBatch.size(), is(2));
+    assertThat(secondBatch.size(), is(1));
 
     Partition altered = alterPartitionCaptor.getValue().get(0);
```
I think to test the batching for `alter_partition` and `setPartitionColumnStatistics` usecase, it will be nice to have more than one existing partitions.

massdosage replied:
Yeah, it feels like we are missing some tests that would have caught the issues you called out above. I'd suggest writing new tests (or modifying existing) and getting them to fail with the current code and then fixing the code and they should then pass.

barnharts4 replied:
hmm, I agree with adding more tests for for `alter_partition` and `setPartitionColumnStatistics` but how would you suggest I get current(old) code to fail?  I guess I could mock a failure of the metastore client if the size of the list was longer than the `partitionBatchSize`, but that seems fairly contrived and not really testing anything except my ability to mock stuff.

barnharts4 replied:
or maybe you didn't mean the issue of overwhelming the HMS with the full list, but rather the issue that I forgot to use `sublist` above?  In that case, yes, good point.

abhimanyugupta07 replied:
Yes. Your last comment is what I meant. Just adding two more existing partitions in the test should give you a batch to trigger the changes you made.

massdosage commented on circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java:
```diff
@@ -30,6 +30,8 @@
 import java.util.Map;
 import java.util.Objects;
 
+import com.google.common.annotations.VisibleForTesting;
```
Minor, but these imports should go along with the other google imports. I've made a mental note to update our hotels-oss-parent with a plugin that automatically does this that we've started using in EG org on GitHub so this is automated by Maven.

barnharts4 replied:
Yeah, I am struggling with the correct IntelliJ settings to match circus train's preferred import layout.  Obviously DSP standards are not matching up well with HCOM in this circumstance. Do you have a doc or a settings file or something so that I can update?

massdosage replied:
This will do it once I've merged and released the parent - https://github.com/HotelsDotCom/circus-train/pull/168/files

I'm getting timeouts publishing to Maven Central right now, they've had lots of scalability issues in the last few months so I'm going to leave it for today and try again tomorrow.

massdosage commented on circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java:
```diff
@@ -89,6 +93,29 @@
     this.housekeepingListener = housekeepingListener;
     replicationMode = tableReplication.getReplicationMode();
     this.tableReplication = tableReplication;
+    this.partitionBatchSize = DEFAULT_PARTITION_BATCH_SIZE;
```
NABD but you could inline this on line 77 and then only have the assignment in the constructor below which actually allows you to override the value.

massdosage commented on circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java:
```diff
@@ -89,6 +93,29 @@
+  }
+
+  /**
+   * Use {@link ReplicaFactory}
```
What does this comment mean?

barnharts4 replied:
It was copied from the other constructor, and it means that this constructor is package-private and to construct this, you need to use `ReplicaFactory`.

massdosage replied:
Ah, OK.

massdosage commented on circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java:
```diff
@@ -175,7 +202,13 @@ public void updateMetadata(
         try {
-          client.add_partitions(partitionsToCreate);
+          int counter = 0;
+          for (List<Partition> sublist : Lists.partition(partitionsToCreate, partitionBatchSize)) {
```
I've never come across `Lists.partition` before, looks good.

massdosage commented on circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java:
```diff
@@ -22,22 +22,29 @@
-import java.util.List;
-import java.util.Map;
-
+import com.google.common.base.Supplier;
```
I know it's not the most readable format but our import ordering is this: https://github.com/ExpediaGroup/eg-oss-parent/blob/master/pom.xml#L102

barnharts4 replied:
ah, thanks, just what I was looking for!

barnharts4 commented:
Updated test to work with any number of added and altered partitions, and any size batch.  Added tests for common edge cases and boundary conditions.  Addressed PR comments.

massdosage commented on circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java:
```diff
@@ -423,41 +466,106 @@ public void alteringExistingPartitionedReplicaTableWithPartitionsSucceeds() thro
+
+  @Test
+  public void alteringExistingPartitionedReplicaTableWithNewPartitionsInBatchesSucceeds_0_0() throws TException, IOException {
+    alterExistingPartitionedReplicaTableWithNewPartitionsInBatches(0,0);
```
For another day but I wonder if we could use JUnit's "parameterised" feature to do this.

barnharts4 replied:
I looked into that, but found referenced to Parameterized not working with `MockitoJUnitRunner` (https://stackoverflow.com/questions/12606148/passing-parameterized-input-using-mockitos) so I didn't pursue further in the interest of time.  But yes, that would be the perfect for this situation.

massdosage replied:
Ah, OK. We should move to JUnit5 at some point as then it can definitely be done, see https://github.com/ExpediaGroup/beekeeper/blob/master/beekeeper-path-scheduler-apiary/src/test/java/com/expediagroup/beekeeper/scheduler/filter/WhitelistedListenerEventFilterTest.java
